action/action.go

```diff
@@ -27,166 +27,310 @@ var (
+		return err
+	}
+	if sealed == nil {
+		return errors.New("nil action to load proto")
+	}
+	*sealed = SealedEnvelope{}
+
+	sealed.srcAddr = pbAct.Sender
+	sealed.srcPubkey = srcPub
+	sealed.signature = make([]byte, len(pbAct.Signature))
+	copy(sealed.signature, pbAct.Signature)
+	sealed.version = pbAct.Version
+	sealed.nonce = pbAct.Nonce
+	sealed.gasLimit = pbAct.GasLimit
+	sealed.gasPrice = &big.Int{}
+	sealed.gasPrice.SetBytes(pbAct.GetGasPrice())
+
+	if pbAct.GetTransfer() != nil {
+		sealed.dstAddr = pbAct.GetTransfer().Recipient
+		act := &Transfer{}
+		if err := act.LoadProto(pbAct.GetTransfer()); err != nil {
+			return err
+		}
+		sealed.payload = act
+	} else if pbAct.GetVote() != nil {
+		sealed.dstAddr = pbAct.GetVote().VoteeAddress
+		act := &Vote{}
+		if err := act.LoadProto(pbAct.GetVote()); err != nil {
+			return err
+		}
+		sealed.payload = act
+	} else if pbAct.GetExecution() != nil {
+		sealed.dstAddr = pbAct.GetExecution().Contract
+		act := &Execution{}
+		if err := act.LoadProto(pbAct.GetExecution()); err != nil {
+			return err
+		}
+		sealed.payload = act
+	} else if pbAct.GetPutBlock() != nil {
+		sealed.dstAddr = pbAct.GetPutBlock().SubChainAddress
+		act := &PutBlock{}
+		if err := act.LoadProto(pbAct.GetPutBlock()); err != nil {
+			return err
+		}
+		sealed.payload = act
+	} else if pbAct.GetStartSubChain() != nil {
+		act := &StartSubChain{}
+		if err := act.LoadProto(pbAct.GetStartSubChain()); err != nil {
+			return err
+		}
+		sealed.payload = act
+	} else if pbAct.GetStopSubChain() != nil {
+		sealed.dstAddr = pbAct.GetStopSubChain().SubChainAddress
+		act := &StopSubChain{}
+		if err := act.LoadProto(pbAct.GetStopSubChain()); err != nil {
+			return err
+		}
+		sealed.payload = act
+	} else if pbAct.GetCreateDeposit() != nil {
+		sealed.dstAddr = pbAct.GetCreateDeposit().Recipient
+		act := &CreateDeposit{}
+		if err := act.LoadProto(pbAct.GetCreateDeposit()); err != nil {
+			return err
+		}
+		sealed.payload = act
+	} else if pbAct.GetSettleDeposit() != nil {
+		sealed.dstAddr = pbAct.GetSettleDeposit().Recipient
+		act := &SettleDeposit{}
+		if err := act.LoadProto(pbAct.GetSettleDeposit()); err != nil {
+			return err
+		}
+		sealed.payload = act
+	} else {
+		return errors.New("no appliable action to handle in action proto")
 	}
+	sealed.payload.SetEnvelopeContext(*sealed)
+	return nil
+}
+
+// Sign signs the action using sender's private key
+func Sign(act Envelope, addr string, sk keypair.PrivateKey) (SealedEnvelope, error) {
+	sealed := SealedEnvelope{Envelope: act}
+
 	// TODO: we should avoid generate public key from private key in each signature
 	pk, err := crypto.EC283.NewPubKey(sk)
 	if err != nil {
-		return errors.Wrapf(err, "error when deriving public key from private key")
+		return sealed, errors.Wrapf(err, "error when deriving public key from private key")
 	}
-	pkHash := keypair.HashPubKey(pk)
-	// TODO: abstract action shouldn't be aware that the playload is the hash of public key
```

Comment:
is it possible to keep this in new code?
it is a safety/security check to make sure the action does come from signer

Reply:
no need to do that check any more. action doesn't tie to an identity til the signing.